book/src/design/circuit/gadgets/ecc/var-base-scalar-mul.md

```diff
@@ -0,0 +1,102 @@
+
+$\begin{array}{cl}
+[2^{254} + (\alpha + t_q)] T &= [2^{254} + (\alpha + t_q) - (2^{254} + t_q)] T \\
+                             &= [\alpha] T
+\end{array}$
+
+provided that $\alpha + t_q \in [0, 2^{n+1})$, i.e. $\alpha < 2^{n+1} - t_q$ which covers the whole range we need because in fact $2^{255} - t_q > q$.
+
+Thus, given a scalar $\alpha$, we witness the boolean decomposition of $k = \alpha + t_q.$ (We use big-endian bit order for convenient input into the variable-base scalar multiplication algorithm.)
+
+$$k = k_{254} \cdot 2^{254} + k_{253} \cdot 2^{253} + \cdots + k_0.$$
+
+## Variable-base scalar multiplication
+We use an optimized double-and-add algorithm is (copied from ["Faster variable-base scalar multiplication in zk-SNARK circuits"](https://github.com/zcash/zcash/issues/3924), with some variable name changes):
+> 
+>     Acc := [2] T
+>     for i from n-1 down to 0 {
+>         P := k_{i+1} ? T : −T
+>         Acc := (Acc + P) + Acc
+>     }
+>     return (k_0 = 0) ? (Acc - T) : Acc
+
+> It remains to check that the x-coordinates of each pair of points to be added are distinct.
+> 
+> When adding points in the large prime-order subgroup, we can rely on Theorem 3 from Appendix C of the [Halo paper](https://eprint.iacr.org/2019/1021.pdf), which says that if we have two such points with nonzero indices wrt a given odd-prime order base, where the indices taken in the range $-(q-1)/2..(q-1)/2$ are distinct disregarding sign, then they have different x-coordinates. This is helpful, because it is easier to reason about the indices of points occurring in the scalar multiplication algorithm than it is to reason about their x-coordinates directly.
+
+> So, the required check is equivalent to saying that the following "indexed version" of the above algorithm never asserts:
+> 
+>     acc := 2
+>     for i from n-1 down to 0 {
+>         p = k_{i+1} ? 1 : −1
+>         assert acc ≠ ± p
+>         assert (acc + p) ≠ acc    // X
+>         acc := (acc + p) + acc
+>         assert 0 < acc ≤ (q-1)/2
+>     }
+>     if k_0 = 0 {
+>         assert acc ≠ 1
+>         acc := acc - 1
+>     }
+
+The maximum value of $acc$ is:
+```
+    <--n 1s--->
+  1011111111111
+= 1100000000000 - 1
+```
+= $2^{n+1} + 2^n - 1$
+
+> The assertion labelled X obviously cannot fail because $u \neq 0$. It is possible to see that acc is monotonically increasing except in the last conditional. It reaches its largest value when $k$ is maximal, i.e. $2^{n+1} + 2^n - 1$.
+
+So to entirely avoid exceptional cases, we would need $2^{n+1} + 2^n - 1 < (q-1)/2$. But we can use $n$ larger by $c$ if the last $c$ iterations use [complete addition](./complete-add.md).
+
+The first $i$ for which the algorithm using **only** incomplete addition fails is going to be $252$, since $2^{252+1} + 2^{252} - 1 > (q - 1)/2$. We need $n = 254$ to make the wraparound technique above work.
+
+> sage: q = 0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001
```

Comment:
```suggestion
```python
sage: q = 0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001
```
and a corresponding ``` at the end (which I can't suggest in a GitHub comment).

Reply:
I think Pygments supports `sage` as a language.

Reply:
I tried that and didn't get any highlighting, so `mdbook` doesn't have that support yet.